homeassistant/components/kostal_plenticore/__init__.py

```diff
@@ -0,0 +1,287 @@
+    CONF_NAME,
+    CONF_PASSWORD,
+    EVENT_HOMEASSISTANT_STOP,
+)
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.device_registry import async_get_registry
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+from homeassistant.util.dt import utcnow
+
+from .const import DOMAIN, SCOPE_PROCESS_DATA, SCOPE_SETTING
+from .sensor import PlenticoreProcessDataSensor
+
+_LOGGER = logging.getLogger(__name__)
+
+CONFIG_SCHEMA = vol.Schema({DOMAIN: vol.Schema({})}, extra=vol.ALLOW_EXTRA)
+
+PLATFORMS = ["sensor"]
+
+SCAN_INTERVAL = timedelta(seconds=10)
+
+
+class PlenticoreApi(DataUpdateCoordinator):
+    """Data Coordinator for fetching all state of the entities.
+
+    Each entity registers on this coordinator as soon as it is registers in hass. The
+    registered entity must provide a scope, module id and data id which is used
+    to poll values batched by scope. The values are returned in the data property
+    in a nested dict structure ([scope][module id][data id]: value).
+
+    Each entity is checked if the current firmware knows about it before polling.
+    If not the entity state is changed to unavailable.
+
+    Entities with scope SCOPE_SETTING are slower much slower polled (5 minutes)
+    than SCOPE_PROCESS_DATA.
+    """
+
+    def __init__(
+        self, hass: HomeAssistant, config: ConfigEntry, logger: logging.Logger
+    ):
+        """Create a new Plenticore Update Coordinator."""
+        super().__init__(
+            hass=hass,
+            logger=logger,
+            name="Plenticore",
+            update_interval=SCAN_INTERVAL,
+            update_method=self._fetch_data,
+        )
+        self.hass = hass
+        self._config = config
+        self._login = False
+        self._registered_entities = []
+        self._available_entities = []
+        self._entities_updated = False
+
+        self._client = PlenticoreApiClient(
+            async_get_clientsession(hass), host=config[CONF_HOST]
+        )
+
+        self._device_info = {
+            "identifiers": {(DOMAIN, config[CONF_HOST])},
+            "name": config[CONF_NAME],
+            "manufacturer": "Kostal",
+        }
+
+        # cache for the fetched process and settings data
+        self._data = {SCOPE_PROCESS_DATA: {}, SCOPE_SETTING: {}}
+
+        # contains all existing module/data ids after login
+        self._existing_data_ids = {SCOPE_PROCESS_DATA: {}, SCOPE_SETTING: {}}
+
+        # last update timestamp of setting values
+        self._last_setting_update = None
+
+    async def logout(self) -> None:
+        """Log the current logged in user out from the API."""
+        if self._login:
+            self._login = False
+            await self._client.logout()
+            _LOGGER.info("Logged out from %s.", self._config[CONF_HOST])
+
+    def register_entity(self, entity: PlenticoreProcessDataSensor) -> None:
```

Comment:
This approach will need to be redesigned. The entity is a platform concern. We're not allowed to expose the entity API outside the entity. Only the core entity platform helper is allowed to access entities from outside.

Probably separate the device api and the entity api. If we need to notify an entity to turn unavailable from outside, we can use our dispatch helper.

But I don't see why we're doing it this way in the first place. The coordinator helper is built to fetch all relevant data and then notify connected listeners (entities) to grab their data from the data cache. We should follow this design.

I would create one coordinator per scope. The coordinator will automatically keep track of connected listeners (entities) and stop the update polling if there are no connected listeners.

The entities can check themselves if their data is available in the fetched data cache and modify their availability accordingly.

Reply:
I am not sure I have understood all the concerns.

I definitely agree with creating one coordinator per scope, so no problem with that.

But I don't understand exactly what you mean by "expose the entity API". Can you be a bit more specific? I am also not sure what you mean by "separate the device api and the entity api".

What I basically want to achieve is this: The coordinator should only query data points that have been added to the `hass` object (i.e. no deactivated entities) and secondly, data points that are not available due to the firmware version should not be queried (`available` property).

Feel free to show me an integration that implements this correctly in a similar way as a blueprint for this integration.

Reply:
Exposing the entity API is allowing something else than the entity to access any of its attributes directly.

The default behavior of the coordinator is to only refresh the data if there are listeners attached. So disabling entities will work as expected.

If we know that an entity should never update due to missing firmware we can either make sure to not add it from the beginning or let it remove itself if it has been added. There's no point in adding an entity that will never be operational.

Reply:
- Use separate coordinator for each scope
- do not expose entity API to coordinator